ostree/ostree_src.go

```diff
@@ -0,0 +1,168 @@
+// +build !containers_image_ostree_stub
+
+package ostree
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/containers/storage/pkg/ioutils"
+	"github.com/ostreedev/ostree-go/pkg/otbuiltin"
+	"github.com/vbatts/tar-split/tar/asm"
+	"github.com/vbatts/tar-split/tar/storage"
+
+	"github.com/containers/image/manifest"
+	"github.com/containers/image/types"
+	"github.com/opencontainers/go-digest"
+	imgspecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+type ostreeImageSource struct {
+	ref    ostreeReference
+	tmpDir string
+}
+
+// newImageSource returns an ImageSource for reading from an existing directory.
+func newImageSource(ctx *types.SystemContext, tmpDir string, ref ostreeReference) (types.ImageSource, error) {
+	return &ostreeImageSource{ref: ref, tmpDir : tmpDir}, nil
+}
+
+// Reference returns the reference used to set up this source.
+func (s *ostreeImageSource) Reference() types.ImageReference {
+	return s.ref
+}
+
+// Close removes resources associated with an initialized ImageSource, if any.
+func (s *ostreeImageSource) Close() error {
+	return nil
+}
+
+func (s *ostreeImageSource) getLayerSize(blob string) (int64, error) {
+	b := fmt.Sprintf("ociimage/%s", blob)
+	// Use golang bindings once they support to read metadata
+	out, err := exec.Command("ostree", "show", "--repo", s.ref.repo, "--print-metadata-key=docker.size", b).CombinedOutput()
+	if err != nil {
+		fmt.Printf("FAIL1\n")
+		return 0, err
+	}
+
+	size, err := strconv.ParseInt(string(out[1:len(out)-2]), 10, 64)
```

Comment:
The `1:len(out)-2` is pretty magic, and fragile. Could this be expressed in safer terms, perhaps splitting trailing '\n' and white space?

(Applies to all instances of that.)

(I guess this would go away with using the bindings directly.)

Reply:
yes that would go away with the bindings.  Anyway, I agree it is quite hard to follow so I've pushed a new version with this part cleaned up.  The `out` string is the output of `g_variant_print` from OSTree plus a `\n`.  With this code I drop the quotes around the JSON file plus the final `\n`.

In the new version, I use `g_variant_parse` so that we don't have to deal how it is quoted.